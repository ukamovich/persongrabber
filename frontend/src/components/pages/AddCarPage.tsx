import React, { useEffect, useState } from "react";
import fetchGrabber, { backendURL } from "../../_helpers/fetchGrabber";

interface CarInterface {
  name: String
  company: String
  production_year: number
  price: number
  owner: string
}

const initialCar: CarInterface = {
  name: "",
  company: "",
  production_year: 0,
  price: 0,
  owner: ""
}

function validateForm(car: CarInterface): boolean {
  for (let [key, value] of Object.entries(car)) {
    if (!value) {
      alert(`${key} cannot be empty!`)
      return false
    }
  }
  return true
}

export function AddCarPage() {

  const [car, setCar] = useState<CarInterface>(initialCar)

  const handleChange = (event: React.ChangeEvent<HTMLInputElement | HTMLSelectElement | HTMLTextAreaElement>) => {
    setCar({ ...car, [event.target.name]: event.target.value })
  }


  const submitForm = (event: React.FormEvent<HTMLFormElement>) => {
    // Preventing the page from reloading
    event.preventDefault();

    if (validateForm(car)) {
      let queryBody = {
        query: `
          mutation {
            createCar(data: {
              name: "${car.name}"
              company: "${car.company}"
              production_year: ${car.production_year}
              price: ${car.price}
              owner: "${car.owner}"
            }) {
              _id
            }
          }
        `
      }
      fetchGrabber(queryBody, backendURL).then(res => {
        alert("Success! New cars id is: " + res.data.createCar._id)
        setCar(initialCar)
      })
    } else {
      console.log("Err!")
    }
  }


  return (

    <form onSubmit={submitForm} style={{textAlign: "center"}}>
      <h3>Add a new car <span className="label label-default"></span></h3>
      <div className="col-md-12">
        <input type="text" className="form-control mt-3" required placeholder="Car model" name="name" onChange={handleChange}/>
      </div>
      <div className="col-md-12">
        <input type="text" className="form-control mt-3" placeholder="Manufactorer" name="company" onChange={handleChange}/>
      </div>
      <div className="col-md-12">
        <input type="number" className="form-control mt-3" placeholder="Production year" name="production_year" onChange={handleChange}/>
      </div>
      <div className="col-md-12">
        <input type="number" className="form-control mt-3" placeholder="Price" name="price" onChange={handleChange}/>
      </div>
      <div className="col-md-12">
        <input type="text" className="form-control mt-3" placeholder="Owner ID" name="owner" onChange={handleChange}/>
      </div>

      <div className="">
        <button type="submit" id="submit" className="btn btn-primary mt-3 mx-3" >Add car</button>
      </div>
    </form>
  );
}