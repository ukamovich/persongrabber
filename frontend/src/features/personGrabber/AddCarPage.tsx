import React, { useEffect, useState } from "react";
import fetchGrabber, { backendURL } from "../../_helpers/fetchGrabber";

interface CarInterface {
  name: string
  company: string
  production_year: number
  price: number
  owner: string
}

const initialCar: CarInterface = {
  name: "",
  company: "",
  production_year: NaN,
  price: NaN,
  owner: ""
}

function AddCarPage() {

  const [car, setCar] = useState<CarInterface>(initialCar)

  const handleChange = (event: React.ChangeEvent<HTMLInputElement | HTMLSelectElement | HTMLTextAreaElement>) => {
    setCar({ ...car, [event.target.name]: event.target.value })
  }


  const submitForm = (event: React.FormEvent<HTMLFormElement>) => {
    
    var forms = document.querySelectorAll('.needs-validation')

    // Preventing the page from reloading
    event.preventDefault();

    Array.prototype.slice.call(forms)
      .forEach(function (form) {
          if (!form.checkValidity()) {
            form.classList.add("was-validated")
            event.stopPropagation();
          }
          else {
            let queryBody = {
              query: `
                mutation {
                  createCar(data: {
                    name: "${car.name}"
                    company: "${car.company}"
                    production_year: ${car.production_year}
                    price: ${car.price}
                    owner: "${car.owner}"
                  }) {
                    _id
                  }
                }
              `
            }
            console.log(queryBody)
            fetchGrabber(queryBody, backendURL).then(res => {
              console.log(res)
              alert("Success! New cars id is: " + res.data.createCar._id)
              setCar(initialCar)
            }).catch(err => {
              alert("You need a persons id!")
            })
          }
        }, false)
  }


  return (

    <form onSubmit={submitForm} noValidate className="needs-validation" style={{textAlign: "center"}}>
      <h3>Add a new car <span className="label label-default"></span></h3>
      <div className="col-md-12">
        <input value={car.name} type="text" required pattern="[a-zA-Z0-9]{2,20}" className="form-control mt-3" placeholder="Car model" name="name" onChange={handleChange}/>
      </div>
      <div className="invalid-feedback">Please enter a valid car model</div>
      <div className="col-md-12">
        <input value={car.company} type="text" required pattern="[a-zA-Z0-9]{2,20}" className="form-control mt-3" placeholder="Manufactorer" name="company" onChange={handleChange}/>
      </div>
      <div className="invalid-feedback">Please pick a car manufactorer</div>
      <div className="col-md-12">
        <input value={isNaN(car.production_year) ? "" : car.production_year} type="number" required pattern="[0-9{4}]" min="1900" max="2022" className="form-control mt-3" placeholder="Production year" name="production_year" onChange={handleChange}/>
      </div>
      <div className="invalid-feedback">Production year on form YYYY</div>
      <div className="col-md-12">
        <input value={isNaN(car.price) ? "" : car.price} type="number" required min="0" className="form-control mt-3" placeholder="Price" name="price" onChange={handleChange}/>
      </div>
      <div className="invalid-feedback">Enter a valid price</div>
      <div className="col-md-12">
        <input value={car.owner} type="text" required className="form-control mt-3" placeholder="Owner ID" name="owner" onChange={handleChange}/>
      </div>
      <div className="invalid-feedback">Enter an owner id</div>
      <div className="">
        <button type="submit" id="submit" className="btn btn-primary mt-3 mx-3" >Add car</button>
      </div>
    </form>
  );
}

export default AddCarPage