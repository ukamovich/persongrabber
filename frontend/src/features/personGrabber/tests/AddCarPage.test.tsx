import AddCarPage from "../AddCarPage";
import { shallow, mount } from "enzyme";
import renderer from "react-test-renderer";

const updateInput = (
  wrapper: any,
  instance: any,
  newValue: any,
  name: string
) => {
  const input = wrapper.find(instance);
  input.simulate("change", {
    target: { value: newValue, name: name },
  });
  return wrapper.find(instance);
};

describe("Testing car creation form", () => {
  test("Renders correctly", () => {
    shallow(<AddCarPage />);
  });

  test("Can input values in fields", () => {
    const wrapper = shallow(<AddCarPage />);
    const nameInput = updateInput(wrapper, '[name="name"]', "Jack", "name");
    expect(nameInput.props().value).toBe("Jack");
    const company = updateInput(
      wrapper,
      '[name="company"]',
      "Volvo",
      "company"
    );
    expect(company.props().value).toBe("Volvo");
    const production_year = updateInput(
      wrapper,
      '[name="production_year"]',
      1999,
      "production_year"
    );
    expect(production_year.props().value).toBe(1999);
  });
});

// Snapshot testing with Jest
describe("Add car page", () => {
  it("renders properly", () => {
    const treeAddCar = renderer.create(<AddCarPage />).toJSON();
    expect(treeAddCar).toMatchSnapshot();
  });
});
