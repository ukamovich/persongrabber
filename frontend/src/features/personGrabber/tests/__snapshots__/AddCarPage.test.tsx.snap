// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add car page renders properly 1`] = `
<form
  className="needs-validation"
  noValidate={true}
  onSubmit={[Function]}
  style={
    Object {
      "textAlign": "center",
    }
  }
>
  <h3>
    Add a new car 
    <span
      className="label label-default"
    />
  </h3>
  <div
    className="col-md-12"
  >
    <input
      className="form-control mt-3"
      id="nameField"
      name="name"
      onChange={[Function]}
      pattern="[a-zA-Z0-9]{2,20}"
      placeholder="Car model"
      required={true}
      type="text"
      value=""
    />
  </div>
  <div
    className="invalid-feedback"
  >
    Please enter a valid car model
  </div>
  <div
    className="col-md-12"
  >
    <input
      className="form-control mt-3"
      name="company"
      onChange={[Function]}
      pattern="[a-zA-Z0-9]{2,20}"
      placeholder="Manufactorer"
      required={true}
      type="text"
      value=""
    />
  </div>
  <div
    className="invalid-feedback"
  >
    Please pick a car manufactorer
  </div>
  <div
    className="col-md-12"
  >
    <input
      className="form-control mt-3"
      max="2022"
      min="1900"
      name="production_year"
      onChange={[Function]}
      pattern="[0-9{4}]"
      placeholder="Production year"
      required={true}
      type="number"
      value=""
    />
  </div>
  <div
    className="invalid-feedback"
  >
    Production year on form YYYY
  </div>
  <div
    className="col-md-12"
  >
    <input
      className="form-control mt-3"
      min="0"
      name="price"
      onChange={[Function]}
      placeholder="Price"
      required={true}
      type="number"
      value=""
    />
  </div>
  <div
    className="invalid-feedback"
  >
    Enter a valid price
  </div>
  <div
    className="col-md-12"
  >
    <input
      className="form-control mt-3"
      name="owner"
      onChange={[Function]}
      placeholder="Owner ID"
      required={true}
      type="text"
      value=""
    />
  </div>
  <div
    className="invalid-feedback"
  >
    Enter an owner id
  </div>
  <button
    className="btn btn-primary mt-3 mx-3"
    id="submit"
    type="submit"
  >
    Add car
  </button>
</form>
`;
